src/Sema.zig

```diff
@@ -12189,7 +12195,33 @@ fn zirReify(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Air.I
             });
             return sema.addType(ty);
         },
-        .ErrorSet => return sema.fail(block, src, "TODO: Sema.zirReify for ErrorSet", .{}),
+        .ErrorSet => {
+            if (union_val.val.isNull())
+                return sema.addType(Type.initTag(.anyerror));
+
+            const payload_val = union_val.val.castTag(.opt_payload).?;
+            const slice_val = payload_val.data.castTag(.slice).?.data;
```

Comment:
This change takes us almost there- only one more thing left to do:

```suggestion
            const payload_val = if (union_val.val.castTag(.opt_payload)) |p| p.data else union_val.val;
            const slice_val = payload_val.castTag(.slice).?.data;
```

Optionals can be represented both ways. This should probably be abstracted into a helper method just like `isNull` so that Sema code does not rely too heavily on checking tag(). (that's not going to block merging this though)

Reply:
Oooooooh. This explains why my "bug fix" in this PR also didn't cause issues unless it was deeply introspected using the raw values/types. Okay. Yes, I understand. I'll poke around and might do the helper too.

Reply:
For some more context: comptime value representation upholds the property that if `coerceInMemoryAllowed` returns `ok` then it is safe to simply change the Type while keeping the Value the same. In other words `coerceInMemoryAllowed` applies to both runtime representation and comptime representation! As a consequence, optional pointer types such as `?*i32` support the same Value representation as non-optional pointers such as `*i32`.

Reply:
Makes sense. Thank you. The change is in 👍